### ISSUE TITLE
Highland v3.0.0

### ISSUE THREAD

**Issue by vqvu** [caolan/highland#179]
This issue tracks any changes we want for v3.0.0. See the 3.0.0 branch.

The biggest change for now is a reimplementation of the Highland engine. See the original PR at #175.

Still to do:

- [ ] See if we can implement `resume` and stream redirects iteratively.
- [ ] Check for memory leaks.
- [ ] Create some benchmarks.
- [ ] "Degenerate tests"

Any other breaking changes? I know of some suggestions in
- #172 - Back propagation of nil, which would require that streams not allow multiple `consumes` without a `fork` (even after the initial `consume` finishes and removes itself).
- #166 - `pipe` passes along errors if piping to a Highland stream.

**Comment by LewisJEllis** [comment]
@vqvu how are iterative resume/redirects going? Is there a way I can be helpful? I was contemplating features/changes for 3.0.0 a bit, but @jeromew wisely suggested that I don't get carried away while the engine is still up in the air.

**Comment by jeromew** [comment]
@LewisJEllis I am not sure @vqvu has had time to look into this over the holiday season. I think that the changes needed to flatten the stack are still up for discovery. Last time @vqvu mentioned it he said he had no idea how convoluted this could get.

I unfortunately could not find time to look deeply at the new branch. If you feel like diving into it, it would be great and @vqvu will certainly appreciate to be able to discuss potential solutions with someone.

I was planning on working on the merge issue https://github.com/caolan/highland/pull/133. @vqvu has rewritten merge in 3.0.0 but, from my understanding, this has nothing to do with the new engine. So I would like to take his new implementation and adapt it to have it meet #133's requirements.

This way we will be able to remove this from the 3.0.0 branch.

After this, we should make sure that 3.0.0 and 2.x are in sync feature wise (I have seen a few commits that need to be checked)

If possible, we should have strictly the same tests between 2.x and 3.0.0

**Comment by vqvu** [comment]
I haven't had time to do any work on Highland for the last few weeks...

@LewisJEllis I'm happy to discuss ideas on how iterative resume/redirect would be done. We also need some sort of a "memory leak checker" and better performance testing. Also the back-prop stuff (this is pretty useful behavior that's not at all implemented) and `pipe` passing errors to Highland streams.

**Comment by jeromew** [comment]
I modified the PR for merge on https://github.com/caolan/highland/pull/133 ; if @vqvu and @LewisJEllis you have time to review it we could remove the merge refactoring from 3.0.0

**Comment by jeromew** [comment]
@LewisJEllis would it make sense to backport ESLint from 3.0.0 to 2.x ? I understand that caolan was ok for the change and don't see a reason to wait for 3.0.0 on this.

**Comment by LewisJEllis** [comment]
Yes, definitely. I originally did only 3.0.0 because I didn't want to make all the silly little changes twice, and because I wasn't sure how much would be done on 2.x in the meantime. It's now evident that 2.x needs it too, and a couple regexes did most of the work anyway, so I'll go ahead and backport it. That should also help avoid silly conflicts when 3.0.0 lands.

Regarding keeping 2.x and 3.0.0 in sync as mentioned in #189 - most things being done on 2.x in the meantime shouldn't cause any major conflicts, so I think an occasional rebase should work just fine. I'll try rebasing 3.0.0 onto master after I backport the linting stuff.

I'm not too familiar with `merge` or the new engine yet, but I'll get up to speed soon so I can take a look at #133 and try to think about the iterative stuff.

**Comment by jeromew** [comment]
ok thanks. We'll have to agree with @vqvu and the 'rebase or not rebase'. I'm all for removing 
 * eslint
 * merge
 * other simple things

from the 3.0.0 potential conflict footprint. I can't say for sure but I think that it will take some time before we get approval from caolan on 3.0.0 so we'd better not freeze 2.x in the meantime and have an easy workflow for merging/rebasing and later have a clean merge from 3.0.0 to master.

**Comment by vqvu** [comment]
@jeromew Why was this issue closed?

As for "rebase vs merge", I suppose it doesn't really matter too much if we don't expect non-collaborators to contribute. Only issue is that if we start getting a large-ish number of commits on 3.0, it'll start getting more annoying and error-prone to do rebases. That said, I haven't actually tried to do the rebase yet, so maybe this is a non-issue.

Agreed that 2.x should still be the branch that gets all back-compat improvements for now.

**Comment by vqvu** [comment]
`throttle`, `debounce`, and `latest` were rewritten to purely use `pull`/`consume` because they depended on private methods and did weird things like override `write` and `pause`.

There's also a few uses of `setImmediate` in the tests that were changed to `_.setImmediate` to get things to work in a browser.

I think these can all be cleanly backported to 2.x.

**Comment by vqvu** [comment]
Same thing with `pipeline`, I think.

**Comment by jeromew** [comment]
@vqvu I backported `throttle`, `debounce` and `latest` + added one missing `_.` before setImmediate in test/tests.

I tried to backport `pipeline` but the tests freeze with the new impl. I did not try to understand why it freezes this is maybe related to the 2.x engine. Do you see a reason why it would freeze ?

**Comment by jeromew** [comment]
@vqvu, @LewisJEllis - I propose that we bump the master version to 2.3.0 and then rebase/merge/recreate the 3.0.0 branch what do you think ?

**Comment by LewisJEllis** [comment]
Sounds good, since most/all of the backporting is done, and 2.x needs a bump anyway.

As for whether to rebase or merge - the difference isn't very significant, but I think it comes down to how we want to land 3.0.0. If we want to have a single giant pull request to merge into master, I think rebase should keep that cleaner (although merge probably also works). If we want to eventually just swap the 3.0.0 branch in for master, merge will theoretically keep more information in the history. I've always used rebase in this sort of situation, so I can't comment as well on the merge option.

**Comment by vqvu** [comment]
Agreed with the 2.x version bump. We can rebase this time to get rid of the backported transforms. I just didn't want to have to rebase every time we add new features to the 2.x branch.

I think I know why `pipeline` doesn't work...probably some difference between the old `_send` and the new `_send`. It doesn't matter too much though; we don't have to backport it.

**Comment by jeromew** [comment]
@caolan, for info, I just bumped the highland npm version to 2.3.0

**Comment by caolan** [comment]
thanks @jeromew, I'm hoping to get some time to look at the engine changes in more depth soon too

**Comment by jeromew** [comment]
@vqvu do you want to try to rebase 3.0.0 or else I might try do it next week.

**Comment by vqvu** [comment]
Yes, I will do it this weekend.

**Comment by vqvu** [comment]
Done. Slightly painful cause of the eslint changes, but it wasn't too bad.

**Comment by jeromew** [comment]
@vqvu thanks. I started digging into it a little trying to see if I could find some performance booster (only using bench.js)

There are 3 things that I tried at this stage:
 - using a derivation of https://github.com/herby/queue to try and have a faster queue. This did not change much on bench.js. hqueue could maybe improve things on a benchmark with lots of async enqueue/dequeue since it tries to re-use memory allocation. In our case in bench.js it does not change much
 - removing the try/catch in `map`. This improves the 1M benchmark. From what I understand, underscore and lodash do not have a unit-based try/catch (I mean around each `f()`call). Not sure what we can/want to do about this
 - inlining the ConsumeStream object. I did not reach an `all tests passing` status because I did this rapidly and was not sure about the teardown but basically I removed the ConsumeStream inheritance by calling directly `new Stream(function() { self.pull(pullCb) }` and defining `pullCb` inside of `consume`. On my setup this gives a perf boost of about 10%.

it seems that in the 2.x version, the 'sync' path is even shorter which is probably why it seems to get a better bench than lodash on the 1M test.

do you have the same results ? (just to make sure that we can compare things on our respective setups)

**Comment by svozza** [comment]
For the try-catch problem - assuming we want to keep them -  would using the trick in the Bluebird performance optimisations document help?

https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#2-unsupported-syntax

>Some constructs are flat out not supported in the optimizing compiler and using such syntax will make the containing function unoptimizable.

>Currently not optimizable:

 > - ...
- Functions that contain a try-catch statement
- Functions that contain a try-finally statement
- ...

>Some of these statements cannot be avoided in production code such as try-finally and try-catch. To use such statements with minimal impact, they must be isolated to a minimal function so that the main code is not affected:

>```javascript
var errorObject = {value: null};
function tryCatch(fn, ctx, args) {
    try {
        return fn.apply(ctx, args);
    }
    catch(e) {
        errorObject.value = e;
        return errorObject;
    }
}

>var result = tryCatch(mightThrow, void 0, [1,2,3]);
//Unambiguously tells whether the call threw
if(result === errorObject) {
    var error = errorObject.value;
}
else {
    //result is the returned value
}
```

Depending on the performance improvement it could be overkill.

**Comment by jeromew** [comment]
@svozza the `map` loop in the 1M case indeed seem to be helped by extruding a function like

````js
function tryCatch(fn, x) {
  try {
        return fn(x);
  }
  catch(e) {
    return new StreamError(e);
  }
}
````
and then doing in the loop

````js
var fnRes;
            fnRes = tryCatch(f, x);
            if (_._isStreamError(fnRes)) {
              push(fnRes);
            } else {
              push(null, fnRes);
            }
````

with this modification + the one on inlining ConsumeStream, the 1M bench.js is on par between lodash and highland on my machine.

the try/catch inside the highland loop has a penalty.

**Comment by jeromew** [comment]
@vqvu do you have a test somewhere regarding the stack exhaustion problem ?

**Comment by vqvu** [comment]
@jeromew There's some test code in the original PR for recursive `next` calls.

Recursive consumes look something like

```javascript
var s = _([1]);

for (var i = 0; i < 100000; i++) {
    s = s.map(function (i) { return i; });
}

s.toArray(_.log);
```

> try-catch

We do want to keep that. This was added intentionally (see #97, #94, and related issues). underscore and lodash don't do error checking because they're synchronous. Since highland is async, we don't want errors just flying off into nowhere.

> perf

I don't really see a benefit to the changes. The code I'm using is at https://github.com/vqvu/highland/tree/3.0.0-perf-test.

Here are my results (node v0.10.18). 

```
# 3.0.0-perf-test
.map(square) x 1,000,000

underscore  #################################################### 122ms
lodash      ################################## 81ms
highland    ############################################################ 143ms


# 3.0.0
.map(square) x 1,000,000

underscore  ######################################################### 131ms
lodash      #################################### 83ms
highland    ############################################################ 139ms
```

That said, I think the move to a specific `tryCatch` function is probably safe enough to make...the question is whether the extra function call is worth it. It's not like the the `consume` callback for `map

**Comment by vqvu** [comment]
Ok, I take back my comment on not seeing a benefit to inlining `ConsumeStream`. I saw some speedup when I modified `bench.js` to add more `map` to the chain.

I added some more aggressive inlining of the sync `consume` code path in my perf test branch.

It inlines all recursive sync calls to `_runGenerator` in `_next_fn` and the call to the generator in `_runGenerator`.

```
.map(square) x 1,000,000

underscore  ############################################################ 141ms
lodash      #################################### 84ms
highland    ########################################## 98ms
```

**Comment by jeromew** [comment]
I cloned your branch and get

````
.map(square) x 1,000,000

underscore  ############################################################ 274ms
lodash      #################################### 162ms
highland    ############################ 126ms
````

the joy of benchmarks. inlining + tryCatch seem to improve things a little.

I have no idea what to do about the bootstrapping of highland at this time. 

````
.map(square) x 10,000
underscore  ######### 1ms
lodash      ######### 1ms
highland    ############################################################ 7ms
````

I am not sure that 2.x does any better but for frequent highland usage on moderate arrays, this can have a big impact. lodash reaches this probably by using the native array.map. I am not sure though how we could 'bridge' the sync/async gap by using array.map when possible.

Anyway we should first see if we want/can flatten the call tree before optimimizing the pull/consume primitives.

On a side note I saw that lodash 3.0.0 has landed and has lazy support - http://filimanjaro.com/blog/2014/introducing-lazy-evaluation/. Maybe there could be a way for the highland engine to defer to lodash when in sync mode. That's just a thought but if the core of highland is really the pull/consume engine that bridges the gap between sync/async, it could be interesting to benefit from the work on lodash for sync optimization..but that's just a tought ;-)

I'll try to understand how the recursive consume could be unrolled so we can discuss this.

**Comment by vqvu** [comment]
Very, very dirty prototype of unlimited `consume`: https://github.com/vqvu/highland/tree/3.0.0-rec-fix. It's really ugly and will be fairly difficult to maintain.

Roughly speaking, the recursive `consume` looks something like this.

```
resume -> rungen -> pull -> ... -> rungen -> _send -> cb -> push -> _writeOutgoing -> _send -> cb -> ...
```

This implementation splits that chain into two parts

```
# Loop is in _runGenerator w/ some coordination from resume. About 5-10% slow down.
resume -> rungen -> pull -> resume -> rungen -> pull
```

and

```
# Loop is purely in _send. About 20% slow down.
_send -> cb -> push -> _writeOutgoing -> _send -> cb -> push -> _writeOutgoing -> _send
```

The second loop is by far the more expensive one, because we have to work around the fact that we have no control over cb.

Unfortunately, the current implementation does come at a roughly 30% cost in the sync benchmark. Still pretty good though; we won't get this for free.

```
.map(square) x 1,000,000

underscore  ############################################################ 130ms
lodash      ###################################### 82ms
highland    ############################################################ 130ms
```

I won't have much time for the rest of the week, but I'll try to take a look at the bluebird wiki. It seems quite useful.

**Comment by vqvu** [comment]
Added infinite `next(xs)` support + some perf improvement.

```
.map(square) x 1,000,000

underscore  ############################################################ 144ms
lodash      ################################### 82ms
highland    ################################################# 116ms
```

**Comment by jeromew** [comment]
@vqvu I fear about waking up this issue because I am not sure how to make this converge with all the parties involved (and mainly @caolan). On the other end, letting this issue unattended feels bad for the evolution of the library if the initial architecture refactoring is a must for solving redirection bugs.

You showed that it is possible to reach the 'flat recursion' objective but I agree that it might be harder to maintain.

Do you think the 'flat recursion' would be a premature optimization or do you feel like the core could stabilize ? Keep in mind a 4.0.0 target with an official extension mechanism so that the main library would only contain the basic building blocks and we could externalize many of the custom transforms that tend to bloat the API.

**Comment by vqvu** [comment]
I think that we need a 3.0.0 release as soon as possible to fix some problems with the library that has popped up. There's backpropagation of nil issue. And the core refactor is needed to fix performance issues with `pull` (see #270) and redirection bugs. We may also want to do some renaming to help out with ramda integration. If "flat recursion" would hold that up, then we don't really need it. There doesn't seem to be many reported issues stemming from this, so I think it's low priority.

I think the important thing is to get 3.0.0 released with the appropriate semantics changes to support the things we want to do. Once we have that, we can easily do everything else as improvements on top.

@caolan I think the ball right now is firmly in your court. If you don't have the time to review the current state of the core engine refactor, please let us know so we can proceed accordingly.

Note that we don't necessarily have to do all of the convoluted things that I did to fix the stack overflow bug. We could just do `setImmediate` everywhere. It's just not ideal for synchronous sources, though there's an argument here that synchronous sources don't matter much since Highland is a library for managing asynchrony.

**Comment by caolan** [comment]
@vqvu your work on this has been excellent, much appreciated! As for the performance improvements, Highland already has lots of complicated performance-related code in 2.x already so I don't see a huge problem in introducing them.

I'm a tentative +1 (since I've not had time to try this out on some real projects yet), but the code looks good and I trust your (and the other collaborators) judgement regarding a release. I'm interested in the ramda integration plans, since that could give us an complimentary tool for sync use cases where performance is critical.

**Comment by jeromew** [comment]
@caolan I want to add that @vqvu is probably the one most knowledgable on the codebase after you. Maybe you would want to consider giving him npm rights (in parallel or instead of me, this is your call depending on how you see things).

You gave me these privileges when I acted as a moderator when @vqvu proposed this huge engine refactoring but I think that I maybe restrain him a little too much on the project wrt the improvements he as made so far (+ I was afk for several weeks and he could not make a release without me).

**Comment by caolan** [comment]
I've added @vqvu as an owner on npm too. You're both welcome to make releases. Please make sure any major or backwards incompatible changes are discussed in an issue first :)

**Comment by vqvu** [comment]
Some updates, thanks to @quarterto for giving me a push on this via #324.

I merged in the latest changes from master and added these features

* Implement `onDestroy`. See #172.
* Disallowed re-consumption of stream (the `take` usecase). Also to support #172.

There's some additional things to do before a proper releas, so I will add some todo tasks in the original task so we can track this stuff.

**Comment by vqvu** [comment]
Ok, I've updated the original post. Anything else that we want to change that we want to change that I've missed.

I know I briefly mentioned in the ramda integration issue that we might want to rename/reorder args for some transforms to make ramda integration easier. Is it worth it to do? Or should we just provide a `ramda-integration` package that people can `use` to override these methods? The other alternative is to provide a `highland-2` package that people `use` to revert to the old transforms. Maybe the latter is better, so new users can get the nicer API while existing users can upgrade incrementally...

See https://github.com/ramda/ramda/issues/1037#issuecomment-94734633 for a detailed description of the differences. The more eggregious ones are

* `reduce`/`scan` has the `iterator` and `memo` arguments reversed.
* `pickBy`'s callback is `f(key, value)` instead of `f(value, key)`.
* `zipAll0` and `zipAll` should really be `zipAll` and `zipEach`, respectively.

Anyway, feedback needed.

**Comment by svozza** [comment]
I'm a +1 on changing those three functions but I don't think we need to go any further than that. 

I really like the idea of a `highland-2` package too.

**Comment by jdalton** [comment]
@caolan
> I'm interested in the ramda integration plans, since that could give us an complimentary tool for sync use cases where performance is critical.

For me it seems like Ramda isn't really about perf critical scenarios opting to prioritize functional philosophy instead. This can be seen in core design choices like Ramda performing more costly deep comparisons for equality checks instead of simpler `===` checks. This compounds for mobile where a recent benchmark on an iPhone 5s took Ramda more than 14 seconds to run an operation on an array 1,000 elements compared to 1 millisecond for a library like Lodash.

**Comment by jdalton** [comment]
FWIW I updated your bench/bench.js to use the latest edge lodash (will be v4), corrected its Underscore usage (I'll send a PR), and enabled a shortcut fusion benchmark:
```bash
.map(square) x 10,000

underscore  ########## 1ms
lodash      ########## 1ms
highland    ############################################################ 6ms


.map(square) x 100,000

underscore  ################### 5ms
lodash      ######## 2ms
highland    ############################################################ 16ms


.map(square) x 1,000,000

underscore  ################################################ 96ms
lodash      ###################### 44ms
highland    ############################################################ 120ms


.map(square).filter(isEven).take(3) x 1,000,000

underscore  ############################################################ 175ms
lodash      # 1ms
highland    #### 9ms
```

**Comment by caolan** [comment]
Thanks @jdalton - I'm continually impressed with lodash :)

**Comment by ericelliott** [comment]
To add a little, those === checks are exactly what you want using immutable data structures, which pair very well with functional programming. :)

**Comment by svozza** [comment]
This is going to sound silly but I want to make the changes to rename `reduce` and the `zip` functions but I don't know how to checkout the `3.0.0` locally and raise a pull request against it.

**Comment by vqvu** [comment]
You probably don't have the 3.0.0 branch in your fork. Do

```
$ git remote add caolan git@github.com:caolan/highland.git
$ git fetch caolan
$ git checkout -b 3.0.0 caolan/3.0.0
$ git push -u origin 3.0.0
```

This will push the 3.0.0 branch to your fork and set it as the upstream for your local 3.0.0 branch. You can then fork a feature branch against like normal. You should be able to [set a base branch](https://help.github.com/articles/using-pull-requests/#changing-the-branch-range-and-destination-repository) when creating your PR.

**Comment by svozza** [comment]
Nice one, I have to do `git push -f origin 3.0.0` to get my fork's branch up to date.

**Comment by svozza** [comment]
@LewisJEllis Do you mind if I port your changes from #166 to the 3.0.0 branch?

**Comment by LewisJEllis** [comment]
I did that in #191 but I think some consensus was reached in #240 that it was no longer needed after some other changes. I'm not too fresh on the details though, so feel free to take another look.

**Comment by svozza** [comment]
Well, I think the reason we didn't do this in 2.x was because it would be a breaking change but I think we still want to do it. #240 was related because `through` had the same behaviour as `pipe` even though it works exclusively with Highland streams so we considered its behaviour a bug.

**Comment by LewisJEllis** [comment]
Right, okay. I do recall the decision not to merge it on 2.x, but I'm otherwise not too sure what the situation is. I left a comment on #191.

**Comment by vqvu** [comment]
@svozza Why is #191 necessary when you can just use `through` and get the same result?

I see a checkbox in the original post about merging in #191 (which I added), but I wasn't thinking about #240 at the time.

**Comment by svozza** [comment]
Well I'm not pushed but it's hardly inconceivable that a user might not read the docs and pipe a Highland stream into another one and if the price of us preventing a strange crash for one of those users is just a simple `if` statement in `pipe` then I don't see why we wouldn't do it.

**Comment by vqvu** [comment]
The standard behavior of `pipe` is to throw if you don't have an error handler attached though, so someone reading the docs would probably assume the standard behavior, right? How about adding a statement in the `pipe` docs pointing to `through` instead?

**Comment by svozza** [comment]
Yep, sounds good to me.

**Comment by vqvu** [comment]
I'll make a PR.

**Comment by LewisJEllis** [comment]
I'm interested in spending some time soon to help push 3.0 out; @vqvu and @quarterto can you give some guidance on what's left to do that I can help with? The todo at the top seems pretty up-to-date, but I haven't been following closely enough to know exactly what each task entails. Happy to help with docs, testing, and any low- to medium-hanging fruit.

Also, is breaking things down into core + `.use`able modules part of the scope of 3.0?

**Comment by vqvu** [comment]
We have:

1) Test for making sure back propagation works for all transforms. Roughly speaking, something along the lines of

```javascript
stream.onDestroy(cb)
    .transform(...)
    .destroy();

// assert that cb was called.
```

for all transforms. This is more of a problem for transforms (listed in the todo) that return a stream created with `this.create` and not `this.consume`. We need a `this.createDownstream` (or equivalent name) that does a `this.create` and binds an `onDestroy` handler like `consume` does.

2) Tests to make sure that a `fork` that is destroyed does not contribute backpressure. That is, assert that the following no longer blocks indefinitely.

```javascript
var s = _([1, 2, 3]);
var s1 = s.fork().take(2);
var s2 = s.fork();

s1.each(_.log);
s2.each(_.log);
// => 1
// => 1
// => 2
// => 2
// => 3
```

3) docs for `use`.

4) a `highland-2` module that restores the behavior of the transforms that changed (i.e., undoes the renames and reargs that we made). I'm not sure exactly how to expose this. Ideally as a separate npm module, but where would we put the repo?

> Also, is breaking things down into a stream engine core + .useable modules part of the scope of 3.0?

I'd like to get this done for 3.0, but we'd want to do this last, since the code movement would make it much harder to merge in the PRs that were applied to 2.x.

**Comment by jeromew** [comment]
@vqvu I saw your mention of 3.0 here - https://github.com/caolan/highland/issues/388#issuecomment-156400784 and realize that bugfixing 2.x engine bugs are a bummer since many things are automatically fixed in the work you did on 3.x.

I think that the engine fixes + the `use` work that you did with @quarterto on https://github.com/caolan/highland/pull/337 are big improvements to the library that we should not hold very much longer.

The main drawback as I understand it would be on performance where we would lose some throughput (I don't have numbers on the last version) but the fixes in the engine + the extensibility would help the community and @vqvu has already shown that there was a way to unroll his engine to gain more speed at the expense of readability of the code.

Regarding speed, I noted one of the remarks of @caolan :

> I'm a tentative +1 (since I've not had time to try this out on some real projects yet), but the code
> looks good and I trust your (and the other collaborators) judgement regarding a release. I'm 
> interested in the ramda integration plans, since that could give us an complimentary tool for sync use > cases where performance is critical.

What is the state-of-affairs is regarding integration of `highland` with other sync libraries that are performance oriented (lodash?). Do we already have an elegant integration these for sync cases when speed is a requirement ?

If we strip down the functions, i would rather call it `highland-core` and `highland-2` but that is just a naming preference.

`highland` could be a version of `highland-core` pre-packaged with basic transforms.

@vqvu how can we help towards the release of 3.0 ?

**Comment by vqvu** [comment]
> I think that the engine fixes + the use work that you did with @quarterto on #337 are big improvements to the library that we should not hold very much longer.

Agreed. The hold up is more me (and I suspect the rest of the collaborators too) not having time to work on this rather than any real blocking issue. @quarterto had a good suggestion to release a 3.0.0-beta1 to npm while we work on finalizing the release, and I think that's a good idea.

> What is the state-of-affairs is regarding integration of highland with other sync libraries that are performance oriented (lodash?). Do we already have an elegant integration these for sync cases when speed is a requirement ?

We currently only have integration with transducers, and while I don't have hard numbers, I suspect using transducers for sync transforms is faster than using highland directly. Transducers could very well be our answer to the sync problem.

I'm not sure about integrations with libraries like lodash. I don't see a way to integrate with them beyond this little snippet.

```javascript
stream.batch(some_reasonable_number)
    .flatMap(function (array) {
        // use lodash here to turn array into result.
        return _(result);
    });
```

At this point, sync performance isn't that bad compared to 2.x, so I'm not too concerned about it. My opinion is that while we should of course care about sync performance the utility of Highland is more about sequencing async computations than raw throughput.

Ramda integration is kind of stalled at the moment, but I think it's more about allowing people to leverage their library of transforms, and especially the FP programming style that comes with it, rather than performance. The integrations that was being proposed did not have to do with performance. As it relates to the 3.0.0 release, I can't think of any further breaking changes that we might need to make to make to support ramda integration, so I think we're safe to proceed here. Unless someone things otherwise?

> If we strip down the functions, i would rather call it highland-core and highland-2 but that is just a naming preference.
> 
> highland could be a version of highland-core pre-packaged with basic transforms.

The `highland-2`  module would be more of a "undo the naming/arg order changes that we did to the transforms". We'd need another one for "non-basic" transforms.

Here's a question for you and the rest of the collaborators. How do we want to handle the multi-module situation? Do we simply package the modules with highland proper? Or is it worth it to create a highland organization to house the different modules in different repos? Bundling everything up would be the simplest, but I think that would defeat the purpose of spliting up the transforms into basic and non-basic packages. If the user is going to download them all anyway, they might as well be allowed to use them by default.

> @vqvu how can we help towards the release of 3.0 ?

I think my previous comment here is a good list of the things still pending. They're all boring documentation/testing tasks, which explains why they've been stalled for so long.

Also, if you or someone else with npm access can release the current 3.0.0 branch as `3.0.0-beta1`, that would be great. The 3.0.0 branch is a little behind master, but it's only lacking the dependency updates we've made within the past week. If possible, it'd also be nice to add a `release:major-beta` so we have it for future uses.

I'd like to refactor `index.js` and `test.js` into multiple files so that they're easier to manage, but that's quite a bit of work and can be done after the release.

**Comment by svozza** [comment]
What exactly is involved in this task:

`Create a test like noValueOnError to test the backpropagation behavior`

I should be able to look at it tomorrow.

**Comment by vqvu** [comment]
Basically, given an infinite stream `s`, if we do

```javascript
var i = 0;
var s = _(function generator(push, next) {
    push(null, i++);
    next();
});

s.onDestroy(function destructor1() {
})
.through(_.someTransform(...))
.onDestroy(function destructor2() {
})
.take(1)
.resume();
```

`destructor2` and `destructor1` should both be called at some point. Furthermore, `generator` should never be called after `destructor2` has been called.

Many transforms get this behavior for free when they use `consume`. However, some (like `sequence`) actually return a new generator stream. These transforms should be using `this.create` to create the new streams, and `create` should take care of setting up the appropriate hooks so that an `onDestroy` on the result stream gets propagated to the parent.

**Comment by svozza** [comment]
Cool. I'll have a go at this tomorrow evening. Btw, I think the idea of creating an organisation and having the various modules in different repos is a great idea.

**Comment by quarterto** [comment]
I did register the org [highlandjs](/highlandjs) a while back...

**Comment by jeromew** [comment]
@caolan would that make sense for you to migrate highland to an highlandjs org ? I don't know if you have followed the work on extensibility that was done by @quarterto and @vqvu in https://github.com/caolan/highland/pull/337 but a lot of modules could exist in this framework.

**Comment by svozza** [comment]
So, I was having a look at this:

> Test that destroyed fork does not contribute backpressure

And it looks like the destroyed `fork` still contributes backpressure. Not quite sure where to look to start fixing it though.

**Comment by vqvu** [comment]
Backpressure for forks is managed by `StreamMultiplexer`. New forks are created using `StreamMultiplexer#newStream`. Destroyed forks are removed via `StreamMultiplexer#removeConsumer`.

When forks want data, they call `StreamMultiplexer#pull` with their id (an integer assigned at creation time) and a callback. This calls `StreamMultiplexer#_resume`. `_resume` will check to see if all forks have requested data. Once this is true, it calls `Stream#pull` on the source and distributes the result to all forks.

Forks may be added or removed while the `Stream#pull` call resolves. If they are added, the result will *not* be distributed. Instead, it's saved in `this._cached_value`.

The problem here is that the multiplexer uses `this._consumers[id] !== undefined` to track that what streams are currently registered ([link](https://github.com/caolan/highland/blob/3.0.0/lib/index.js#L4050)). However, the `this._consumers` object [gets reset every time a value is emitted](https://github.com/caolan/highland/blob/3.0.0/lib/index.js#L3982), which wipes away that information.

Fix in #411.

**Comment by svozza** [comment]
Makes sense. Thanks for the explanation!
